### ISSUE TITLE
View not Resolving and Error Message

### ISSUE THREAD

**Issue by RickStrahl** [aspnet/Mvc#4810]
I'm going through an exercise of building up an ASP.NET Core app up from scratch. From empty CommandLine scaffold through middleware handler, middleware, MVC API, and MVC, just to see all the pieces needed for each step of the way.  It's all working all the way except to the MVC pieces which don't seem to be able to resolve the View location.

I'm getting:

>An unhandled exception occurred while processing the request.
>
>InvalidOperationException: The view 'Goodbye' was not found. The following locations were searched:
>/Views/Helloworld/Goodbye.cshtml
>/Views/Shared/Goodbye.cshtml

Now the view exists in the `/views/helloworld folder, but no matter what I try it's not picking up the view. I've tried using the default view resolution (Controller/Method), explicit path ~/Views/Helloworld/Goodbye and different path formatting, but none of it seems to work. I can't get past the above error **on this project** (it works in another project). I've compared startup config for both projects and I don't see what is different in terms of MVC and path configuration.

One issue I have here is that that the (debug) error message doesn't point at the the physical path where it's searching. That would be really helpful - I'm guessing somehow the application is not using the correct base path for the application. I think I'm setting it in the configuration.

Here's the complete applet:

```csharp
using System;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Hosting.Internal;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.AspNetCore.Routing;
using Microsoft.Extensions.Configuration;

namespace ConsoleApplication
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var host = new WebHostBuilder()
                .UseKestrel()
                .UseStartup<Startup>()
                .Build();

            host.Run();
        }
    }

    public class Startup
    {
        private IHostingEnvironment HostingEnvironment;
        IConfigurationRoot Configuration { get; }

        public Startup(IHostingEnvironment env)
        {
            HostingEnvironment = env;

            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath);

            Configuration = builder.Build();
        }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc();
        }

        public void Configure(IApplicationBuilder app)
        {

            app.UseDeveloperExceptionPage();

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
            
            // this becomes the last handler in the chain
            app.Run(async (context) =>
            {
                await context.Response.WriteAsync(
                    "Hello World. The Time is ss: " +
                    DateTime.Now);
            });
        }

    }

    public class HelloworldController : Controller
    {
        private IHostingEnvironment Environment;

        public HelloworldController(IHostingEnvironment env)
        {
            Environment = env;
        }

        [HttpGet("api/helloworld")]
        public object Helloworld()
        {
            return new
            {
                message = "Hello World",
                time = DateTime.Now
            };
        }

        [HttpGet("goodbye")]
        public ActionResult Goodbye()
        {
            //ViewData["Message"] = "Hello world!";
            //ViewBag["Time"] = DateTime.Now;

            return View();
            //return View("~/helloworld/goodbye.cshtml");  // exists!
        }
    }
}
```

**Suggestion**: When Views don't resolve, please provide a physical path where the ViewResolver is looking.

**Comment by RickStrahl** [comment]
Ok I found the problem. I needed `.UseContentRoot()`:

```
public static void Main(string[] args)
        {
            var host = new WebHostBuilder()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseKestrel()
                .UseStartup<Startup>()                
                .Build();

            host.Run();
        }
```

and `preserveCompilationContext` in the build options in `project.json`:

```
  "buildOptions": {
    "emitEntryPoint": true,
    "preserveCompilationContext": true
  },
```

Oddly, my other sample project doesn't use either of these and it works - shrug.

**Comment by davidfowl** [comment]
> Oddly, my other sample project doesn't use either of these and it works - shrug.

It *does* need them 😄

**Comment by RickStrahl** [comment]
Yeah but it must be getting the base from something else as that sample is not using `UseContentRoot()` in main.

I wonder though - why is this really needed? You're launching a Console application, and the folder that it's running out of is known at the time of startup and the folder also shouldn't really change unless the application explicitly changes it. So why couldn't this value be inferred by default as a convention?

**Comment by davidfowl** [comment]
> So why couldn't this value be inferred by default as a convention?

The default is looking for views, config files, and the web root next to the .exe. Those are not copied on build because it makes for a poor development experience (build step required to see javascript/razor/config changes). The middle ground is using the working directory (which is usually ok) from within the application itself.

**Comment by RickStrahl** [comment]
Not sure I understand. If I can do this manually in Main():

```
 .UseContentRoot(Directory.GetCurrentDirectory())
```
to set the directory, so can the framework. It just seems 99% of the time this is what you want, so this probably shouldn't be a required setting. WebHostBuilder could default to this value easily enough, no?

I ended up here because I started from a non-ASP.NET project then added ASP.NET features and the default template for .NET Core project doesn't add that setting. It's easy to miss.

When the views were failing there's also no way to see what the actual resolved path is and that would be really helpful to track this down. I do wonder what the actual path was since it wasn't pointing at the right path. What other folder could it possibly be pointing at when I'm running from the command line? What other 'default' path is there?
